contrib/vcloud/benchmarkclient_executor.py

```diff
@@ -0,0 +1,449 @@
+# This file is part of BenchExec, a framework for reliable benchmarking:
+# https://github.com/sosy-lab/benchexec
+#
+# SPDX-FileCopyrightText: 2007-2020 Dirk Beyer <https://www.sosy-lab.org>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+import json
+import logging
+import os
+import shutil
+import subprocess
+
+import benchexec.tooladapter
+import benchexec.util
+from . import util as util
+
+sys.dont_write_bytecode = True  # prevent creation of .pyc files
+
+DEFAULT_CLOUD_TIMELIMIT = 300  # s
+
+DEFAULT_CLOUD_MEMORY_REQUIREMENT = 7000000000  # 7 GB
+DEFAULT_CLOUD_CPUCORE_REQUIREMENT = 2  # one core with hyperthreading
+DEFAULT_CLOUD_CPUMODEL_REQUIREMENT = ""  # empty string matches every model
+
+STOPPED_BY_INTERRUPT = False
+
+# Assuming that benchmark.py is executed from the contrib folder.
+_ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__)))
+
+# Check if benchmark.py is executed from the contrib folder,
+# otherwise it is being used from the scripts folder in CPAchecker.
+if not os.path.isfile(os.path.join(_ROOT_DIR, "build.xml")):
+    _ROOT_DIR = os.path.abspath(
+        os.path.join(os.path.dirname(__file__), os.pardir, os.pardir)
+    )
+
+# Keeping the variable name as before but adding a noqa
+_justReprocessResults = False  # noqa N816
+
+
+def init(config, benchmark):
+    global _justReprocessResults
+    _justReprocessResults = config.reprocessResults
+    tool_locator = benchexec.tooladapter.create_tool_locator(config)
+    benchmark.executable = benchmark.tool.executable(tool_locator)
+    benchmark.tool_version = benchmark.tool.version(benchmark.executable)
+    environment = benchmark.environment()
+    if environment.get("keepEnv", None) or environment.get("additionalEnv", None):
+        sys.exit(
+            "Unsupported environment configuration in tool-info module, "
+            'only "newEnv" is supported by VerifierCloud'
+        )
+
+
+def get_system_info():
+    return None
+
+
+def execute_benchmark(benchmark, output_handler):
+    if not _justReprocessResults:
+        # build input for cloud
+        (cloudInput, numberOfRuns) = getCloudInput(benchmark)
+        if benchmark.config.debug:
+            cloudInputFile = os.path.join(benchmark.log_folder, "cloudInput.txt")
+            benchexec.util.write_file(cloudInput, cloudInputFile)
+            output_handler.all_created_files.add(cloudInputFile)
+        meta_information = json.dumps(
+            {
+                "tool": {
+                    "name": benchmark.tool_name,
+                    "revision": benchmark.tool_version,
+                    "benchexec-module": benchmark.tool_module,
+                },
+                "benchmark": benchmark.name,
+                "timestamp": benchmark.instance,
+                "generator": "benchmark.vcloud.py",
+            }
+        )
+
+        if not os.getenv("TAKE_LOCAL_VCLOUD_JAR"):
```

Comment:
Having an environment variable like this is quite an obscure UI. Is it really necessary or can we detect somehow whether to use Ivy? If not, I would suggest a parameter like `--vcloudJar` that users could use for passing the path to the JAR, and if it is not set, the script downloads it using Ivy.

Reply:
I have removed it for the moment.
I would prefer to go ahead with this PR first and then add this feature when the need arise.

Reply:
Please do not check in JAR files. To not needlessly increase the history size I would prefer if the last commit could be edited to not contain the JAR file and then be force-pushed.

Reply:
Futhermore, the code that you said is removed is still there?

We should also think about how to make this work in way that allows easy sharing of this file with the CPAchecker repository. I would suggest to move the code that is responsible for retrieving the JAR file to `vcloud-benchmark.py` and just expect a path to the JAR to be passed to `benchmarkclient_executor.py` somehow. Then `vcloud-benchmark.py` could retrieve the JAR via Ivy or use a user-supplied one, and CPAchecker's `benchmark.py` could use the existing `ant`-based setup.

Reply:
> Please do not check in JAR files. To not needlessly increase the history size I would prefer if the last commit could be edited to not contain the JAR file and then be force-pushed.

Which jar files? We would need `ivy.jar`.

Reply:
> Futhermore, the code that you said is removed is still there?
It is there now. I had forgotten to commit.

> We should also think about how to make this work in way that allows easy sharing of this file with the CPAchecker repository. I would suggest to move the code that is responsible for retrieving the JAR file to `vcloud-benchmark.py` and just expect a path to the JAR to be passed to `benchmarkclient_executor.py` somehow. Then `vcloud-benchmark.py` could retrieve the JAR via Ivy or use a user-supplied one, and CPAchecker's `benchmark.py` could use the existing `ant`-based setup.

Why not just use the same thing for both, i.e., remove the `resolve-benchmark-dependencies` from CPAchecker's `build.xml`?

Reply:
> > Please do not check in JAR files. To not needlessly increase the history size I would prefer if the last commit could be edited to not contain the JAR file and then be force-pushed.
> 
> Which jar files? We would need `ivy.jar`.

We should just download it like the `ant` file did (whether we use `ant` or not does not change anything regarding the arguments of having it checked in or not).

[`urllib.request.urlretrieve`](https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve) should be usable here, it is the shortest way I know to download a file in Python.

Reply:
> Why not just use the same thing for both, i.e., remove the `resolve-benchmark-dependencies` from CPAchecker's `build.xml`?

Hmm, interesting idea. It would have the advantage of more clearly separating the benchmarking dependencies from CPAchecker's actual dependencies (by removing the former from `ivy.xml`). On the other hand there would be two `ivy.jar` lying around, and JAR files would probably be lying somewhere below `scripts` instead of `lib`. But we can try out and see how it looks?

Reply:
>urllib.request.urlretrieve should be usable here, it is the shortest way I know to download a file in Python.

from here: https://www.sosy-lab.org/ivy/org.apache.ivy/ivy/ivy-2.5.0.jar ?

Reply:
Yes.

Thinking about it, it would be good to use the name `ivy-2.5.0.jar` for the local file. Then we can skip retrieving the file if it already exists, and still will be able to update the used Ivy version by changing the file name. If we use `ivy.jar` we would either have to download every time or there would be update problems.

Reply:
done